File: bindings/nodejs/src/validation.ts
/**
 * MAINFRAME Node.js Bindings - Validation Module
 *
 * TypeScript wrappers for MAINFRAME validation and sanitization functions.
 * Calls validation.sh functions: validate_email, validate_url, sanitize_html, etc.
 */

import { callFunction, callFunctionRaw } from "./core.js";

// =============================================================================
// Type Validation
// =============================================================================

/**
 * Validate an integer with optional range
 *
 * @example
 * validateInt("42")           // true
 * validateInt("42", 0, 100)   // true
 * validateInt("150", 0, 100)  // false
 */
export function validateInt(
  value: string | number,
  min?: number,
  max?: number,
): boolean {
  const args: (string | number)[] = [String(value)];
  if (min !== undefined) args.push(min);
  if (max !== undefined) args.push(max);

  const result = callFunction<string>("validate_int", args);
  return result.success;
}

/**
 * Validate a floating point number
 *
 * @example
 * validateFloat("12.34")  // true
 * validateFloat("1e10")   // true
 * validateFloat("abc")    // false
 */
export function validateFloat(value: string | number): boolean {
  const result = callFunction<string>("validate_float", [String(value)]);
  return result.success;
}

/**
 * Validate a boolean value
 * Accepts: true/false/yes/no/1/0/on/off (case insensitive)
 *
 * @example
 * validateBool("true")   // true
 * validateBool("yes")    // true
 * validateBool("maybe")  // false
 */
export function validateBool(value: string): boolean {
  const result = callFunction<string>("validate_bool", [value]);
  return result.success;
}

/**
 * Validate UUID format (v1-v5)
 *
 * @example
 * validateUuid("550e8400-e29b-41d4-a716-446655440000")  // true
 * validateUuid("not-a-uuid")                            // false
 */
export function validateUuid(value: string): boolean {
  const result = callFunction<string>("validate_uuid", [value]);
  return result.success;
}

/**
 * Validate hexadecimal string
 *
 * @example
 * validateHex("deadbeef")      // true
 * validateHex("deadbeef", 8)   // true (exact length)
 * validateHex("xyz")           // false
 */
export function validateHex(value: string, length?: number): boolean {
  const args: (string | number)[] = [value];
  if (length !== undefined) args.push(length);

  const result = callFunction<string>("validate_hex", args);
  return result.success;
}

// =============================================================================
// Format Validation
// =============================================================================

/**
 * Validate email address (RFC 5322 simplified)
 *
 * @example
 * validateEmail("user@example.com")  // true
 * validateEmail("invalid")           // false
 */
export function validateEmail(email: string): boolean {
  const result = callFunction<string>("validate_email", [email]);
  return result.success;
}

/**
 * Validate URL format
 *
 * @example
 * validateUrl("https://example.com")           // true
 * validateUrl("ftp://files.com", "ftp,http")   // true
 * validateUrl("not-a-url")                     // false
 */
export function validateUrl(url: string, schemes?: string): boolean {
  const args = [url];
  if (schemes) args.push(schemes);

  const result = callFunction<string>("validate_url", args);
  return result.success;
}

/**
 * Validate domain name format
 *
 * @example
 * validateDomain("example.com")  // true
 * validateDomain("localhost")    // false (no TLD)
 */
export function validateDomain(domain: string): boolean {
  const result = callFunction<string>("validate_domain", [domain]);
  return result.success;
}

/**
 * Validate IPv4 address
 *
 * @example
 * validateIpv4("192.168.1.1")  // true
 * validateIpv4("999.1.1.1")    // false
 */
export function validateIpv4(ip: string): boolean {
  const result = callFunction<string>("validate_ipv4", [ip]);
  return result.success;
}

/**
 * Validate IPv6 address
 *
 * @example
 * validateIpv6("::1")                           // true
 * validateIpv6("2001:db8::1")                   // true
 * validateIpv6("2001:db8:85a3::8a2e:370:7334")  // true
 */
export function validateIpv6(ip: string): boolean {
  const result = callFunction<string>("validate_ipv6", [ip]);
  return result.success;
}

/**
 * Validate date format YYYY-MM-DD
 *
 * @example
 * validateDate("2024-01-15")  // true
 * validateDate("2024-02-30")  // false (invalid day)
 */
export function validateDate(date: string): boolean {
  const result = callFunction<string>("validate_date", [date]);
  return result.success;
}

/**
 * Validate time format HH:MM:SS
 *
 * @example
 * validateTime("12:30:00")  // true
 * validateTime("25:00:00")  // false
 */
export function validateTime(time: string): boolean {
  const result = callFunction<string>("validate_time", [time]);
  return result.success;
}

/**
 * Validate semantic version
 *
 * @example
 * validateSemver("1.2.3")        // true
 * validateSemver("v1.2.3")       // true
 * validateSemver("1.2.3-beta")   // true
 * validateSemver("1.2")          // false
 */
export function validateSemver(version: string): boolean {
  const result = callFunction<string>("validate_semver", [version]);
  return result.success;
}

// =============================================================================
// Path Validation
// =============================================================================

/**
 * Validate path exists with optional type check
 *
 * @example
 * validatePath("/etc/hosts", "file")  // true
 * validatePath("/tmp", "dir")         // true
 */
export function validatePath(
  path: string,
  type?: "file" | "dir" | "any",
): boolean {
  const args = [path];
  if (type) args.push(type);

  const result = callFunction<string>("validate_path", args);
  return result.success;
}

/**
 * Validate path is safe (no traversal attacks)
 *
 * @example
 * validatePathSafe("./file.txt")           // true
 * validatePathSafe("../../../etc/passwd")  // false
 */
export function validatePathSafe(path: string, baseDir?: string): boolean {
  const args = [path];
  if (baseDir) args.push(baseDir);

  const result = callFunction<string>("validate_path_safe", args);
  return result.success;
}

/**
 * Validate filename (no path components)
 *
 * @example
 * validateFilename("file.txt")      // true
 * validateFilename("../file.txt")   // false
 * validateFilename("-file.txt")     // false (starts with dash)
 */
export function validateFilename(name: string): boolean {
  const result = callFunction<string>("validate_filename", [name]);
  return result.success;
}

/**
 * Validate path contains only safe characters
 *
 * @example
 * validatePathChars("/home/user/file.txt")  // true
 * validatePathChars("/home/user/file;rm")   // false
 */
export function validatePathChars(path: string): boolean {
  const result = callFunction<string>("validate_path_chars", [path]);
  return result.success;
}

// =============================================================================
// Sanitization
// =============================================================================

/**
 * Sanitize value for safe shell argument
 *
 * @example
 * sanitizeShellArg("hello world")  // 'hello world' (or properly escaped)
 * sanitizeShellArg("$(whoami)")    // \$\(whoami\)
 */
export function sanitizeShellArg(value: string): string {
  return callFunctionRaw("sanitize_shell_arg", [value]);
}

/**
 * Sanitize filename by removing dangerous characters
 *
 * @example
 * sanitizeFilename("my/file.txt")      // my_file.txt
 * sanitizeFilename("file;rm -rf /")    // file_rm_-rf__
 */
export function sanitizeFilename(name: string, replacement?: string): string {
  const args = [name];
  if (replacement) args.push(replacement);

  return callFunctionRaw("sanitize_filename", args);
}

/**
 * Sanitize for SQL (basic escaping)
 * WARNING: Prefer parameterized queries in production
 *
 * @example
 * sanitizeSql("O'Brien")  // O''Brien
 */
export function sanitizeSql(value: string): string {
  return callFunctionRaw("sanitize_sql", [value]);
}

/**
 * Sanitize for HTML (entity escaping)
 *
 * @example
 * sanitizeHtml('<script>alert("xss")</script>')
 * // &lt;script&gt;alert(&quot;xss&quot;)&lt;/script&gt;
 */
export function sanitizeHtml(value: string): string {
  return callFunctionRaw("sanitize_html", [value]);
}

/**
 * Sanitize for JSON string (escape special chars)
 *
 * @example
 * sanitizeJson('hello\nworld')  // hello\\nworld
 */
export function sanitizeJson(value: string): string {
  return callFunctionRaw("sanitize_json", [value]);
}

// =============================================================================
// Complex Validation
// =============================================================================

/**
 * Validate value against regex pattern
 *
 * @example
 * validateRegex("abc123", "^[a-z0-9]+$")  // true
 */
export function validateRegex(value: string, pattern: string): boolean {
  const result = callFunction<string>("validate_regex", [value, pattern]);
  return result.success;
}

/**
 * Validate string length
 *
 * @example
 * validateLength("hello", 1, 10)  // true
 * validateLength("hi", 5)         // false (min 5)
 */
export function validateLength(
  value: string,
  min?: number,
  max?: number,
): boolean {
  const args: (string | number)[] = [value];
  // Always push min if max is defined (to maintain positional args)
  if (min !== undefined || max !== undefined) {
    args.push(min !== undefined ? min : "");
  }
  if (max !== undefined) args.push(max);

  const result = callFunction<string>("validate_length", args);
  return result.success;
}

/**
 * Validate value is in list (enum)
 *
 * @example
 * validateEnum("red", ["red", "green", "blue"])  // true
 * validateEnum("yellow", ["red", "green"])       // false
 */
export function validateEnum(value: string, options: string[]): boolean {
  const result = callFunction<string>("validate_enum", [value, ...options]);
  return result.success;
}

// =============================================================================
// Convenience Validators
// =============================================================================

/** Validation result with error message */
export interface ValidationResult {
  valid: boolean;
  error?: string;
}

/**
 * Validate with detailed error messages
 */
export const validators = {
  email: (value: string): ValidationResult => ({
    valid: validateEmail(value),
    error: validateEmail(value) ? undefined : "Invalid email format",
  }),

  url: (value: string): ValidationResult => ({
    valid: validateUrl(value),
    error: validateUrl(value) ? undefined : "Invalid URL format",
  }),

  uuid: (value: string): ValidationResult => ({
    valid: validateUuid(value),
    error: validateUuid(value) ? undefined : "Invalid UUID format",
  }),

  ipv4: (value: string): ValidationResult => ({
    valid: validateIpv4(value),
    error: validateIpv4(value) ? undefined : "Invalid IPv4 address",
  }),

  ipv6: (value: string): ValidationResult => ({
    valid: validateIpv6(value),
    error: validateIpv6(value) ? undefined : "Invalid IPv6 address",
  }),

  date: (value: string): ValidationResult => ({
    valid: validateDate(value),
    error: validateDate(value)
      ? undefined
      : "Invalid date format (expected YYYY-MM-DD)",
  }),

  semver: (value: string): ValidationResult => ({
    valid: validateSemver(value),
    error: validateSemver(value) ? undefined : "Invalid semantic version",
  }),
};

File: bindings/python/mainframe_bash/validation.py
"""
MAINFRAME Validation Functions - Python wrappers for input validation.

Wraps validation.sh functions: validate_*, sanitize_*, etc.
"""

from .core import MainframeFunctionError
from .core import _fixed_convenience_call as call_function


def validate_email(email: str) -> bool:
    """
    Validate email address format (RFC 5322 simplified).

    Args:
        email: Email address to validate.

    Returns:
        True if valid, False otherwise.

    Example:
        validate_email("test@example.com")  # True
        validate_email("invalid")           # False
    """
    _, code = call_function("validate_email", email)
    return code == 0


def validate_url(url: str, schemes: str = "http,https") -> bool:
    """
    Validate URL format.

    Args:
        url: URL to validate.
        schemes: Comma-separated list of allowed schemes.

    Returns:
        True if valid, False otherwise.

    Example:
        validate_url("https://example.com")  # True
        validate_url("ftp://example.com")    # False (default schemes)
        validate_url("ftp://x.com", "ftp")   # True
    """
    _, code = call_function("validate_url", url, schemes)
    return code == 0


def validate_uuid(value: str) -> bool:
    """
    Validate UUID format (v1-v5).

    Args:
        value: String to validate.

    Returns:
        True if valid UUID format, False otherwise.

    Example:
        validate_uuid("550e8400-e29b-41d4-a716-446655440000")  # True
        validate_uuid("not-a-uuid")                            # False
    """
    _, code = call_function("validate_uuid", value)
    return code == 0


def validate_int(
    value: str | int,
    min_val: int | None = None,
    max_val: int | None = None,
) -> bool:
    """
    Validate integer with optional range.

    Args:
        value: Value to validate.
        min_val: Minimum allowed value (optional).
        max_val: Maximum allowed value (optional).

    Returns:
        True if valid integer in range, False otherwise.

    Example:
        validate_int("42")           # True
        validate_int("42", 0, 100)   # True
        validate_int("42", 50, 100)  # False
    """
    args = [str(value)]
    if min_val is not None:
        args.append(str(min_val))
    if max_val is not None:
        if min_val is None:
            args.append("")  # placeholder for min
        args.append(str(max_val))

    _, code = call_function("validate_int", *args)
    return code == 0


def validate_float(value: str | float) -> bool:
    """
    Validate floating point number.

    Args:
        value: Value to validate.

    Returns:
        True if valid float, False otherwise.

    Example:
        validate_float("3.14")    # True
        validate_float("1e-10")   # True
        validate_float("abc")     # False
    """
    _, code = call_function("validate_float", str(value))
    return code == 0


def validate_bool(value: str | bool) -> bool:
    """
    Validate boolean value.

    Accepts: true/false/yes/no/1/0/on/off (case insensitive).

    Args:
        value: Value to validate.

    Returns:
        True if valid boolean representation, False otherwise.

    Example:
        validate_bool("true")   # True
        validate_bool("yes")    # True
        validate_bool("maybe")  # False
    """
    if isinstance(value, bool):
        return True
    _, code = call_function("validate_bool", str(value))
    return code == 0


def validate_date(date: str) -> bool:
    """
    Validate date format YYYY-MM-DD.

    Also validates the date is actually valid (correct days in month, leap years).

    Args:
        date: Date string to validate.

    Returns:
        True if valid date, False otherwise.

    Example:
        validate_date("2024-01-15")  # True
        validate_date("2024-02-30")  # False (invalid day)
    """
    _, code = call_function("validate_date", date)
    return code == 0


def validate_time(time: str) -> bool:
    """
    Validate time format HH:MM:SS.

    Args:
        time: Time string to validate.

    Returns:
        True if valid time, False otherwise.

    Example:
        validate_time("14:30:00")  # True
        validate_time("25:00:00")  # False
    """
    _, code = call_function("validate_time", time)
    return code == 0


def validate_semver(version: str) -> bool:
    """
    Validate semantic version format.

    Accepts versions with optional 'v' prefix and pre-release/build metadata.

    Args:
        version: Version string to validate.

    Returns:
        True if valid semver, False otherwise.

    Example:
        validate_semver("1.2.3")            # True
        validate_semver("v1.2.3-beta+build") # True
        validate_semver("1.2")               # False
    """
    _, code = call_function("validate_semver", version)
    return code == 0


def validate_ipv4(ip: str) -> bool:
    """
    Validate IPv4 address format.

    Args:
        ip: IP address to validate.

    Returns:
        True if valid IPv4 address, False otherwise.

    Example:
        validate_ipv4("192.168.1.1")  # True
        validate_ipv4("256.1.1.1")    # False
    """
    _, code = call_function("validate_ipv4", ip)
    return code == 0


def validate_ipv6(ip: str) -> bool:
    """
    Validate IPv6 address format.

    Args:
        ip: IP address to validate.

    Returns:
        True if valid IPv6 address, False otherwise.

    Example:
        validate_ipv6("2001:db8::1")  # True
        validate_ipv6("invalid")      # False
    """
    _, code = call_function("validate_ipv6", ip)
    return code == 0


def validate_domain(domain: str) -> bool:
    """
    Validate domain name format.

    Args:
        domain: Domain name to validate.

    Returns:
        True if valid domain, False otherwise.

    Example:
        validate_domain("example.com")  # True
        validate_domain("-invalid.com") # False
    """
    _, code = call_function("validate_domain", domain)
    return code == 0


def validate_hex(value: str, length: int | None = None) -> bool:
    """
    Validate hexadecimal string.

    Args:
        value: String to validate.
        length: Expected length (optional).

    Returns:
        True if valid hex string, False otherwise.

    Example:
        validate_hex("deadbeef")       # True
        validate_hex("abcd", 4)        # True
        validate_hex("abc", 4)         # False (wrong length)
    """
    args = [value]
    if length is not None:
        args.append(str(length))

    _, code = call_function("validate_hex", *args)
    return code == 0


def validate_path(path: str, path_type: str = "any") -> bool:
    """
    Validate path exists with optional type check.

    Args:
        path: Path to validate.
        path_type: "file", "dir", or "any" (default).

    Returns:
        True if path exists and matches type, False otherwise.

    Example:
        validate_path("/etc/passwd", "file")  # True
        validate_path("/etc", "dir")          # True
    """
    _, code = call_function("validate_path", path, path_type)
    return code == 0


def validate_path_safe(path: str, base_dir: str | None = None) -> bool:
    """
    Validate path is safe (no traversal attacks).

    Prevents: ../, symlink escapes, null bytes.

    Args:
        path: Path to validate.
        base_dir: Base directory path must stay within (optional).

    Returns:
        True if path is safe, False otherwise.

    Example:
        validate_path_safe("/app/data/file.txt", "/app/data")  # True
        validate_path_safe("../etc/passwd", "/app/data")       # False
    """
    args = [path]
    if base_dir is not None:
        args.append(base_dir)

    _, code = call_function("validate_path_safe", *args)
    return code == 0


def validate_filename(name: str) -> bool:
    """
    Validate filename (no path components).

    Rejects path separators, special names (. / ..), and leading dashes.

    Args:
        name: Filename to validate.

    Returns:
        True if valid filename, False otherwise.

    Example:
        validate_filename("report.pdf")  # True
        validate_filename("../file")     # False
        validate_filename("-flag")       # False
    """
    _, code = call_function("validate_filename", name)
    return code == 0


def validate_length(
    value: str,
    min_len: int | None = None,
    max_len: int | None = None,
) -> bool:
    """
    Validate string length.

    Args:
        value: String to validate.
        min_len: Minimum length (optional).
        max_len: Maximum length (optional).

    Returns:
        True if length is within bounds, False otherwise.

    Example:
        validate_length("hello", 1, 10)   # True
        validate_length("hi", 5, 10)      # False
    """
    args = [value]
    if min_len is not None:
        args.append(str(min_len))
    if max_len is not None:
        if min_len is None:
            args.append("")  # placeholder
        args.append(str(max_len))

    _, code = call_function("validate_length", *args)
    return code == 0


def sanitize_html(value: str) -> str:
    """
    Sanitize string for safe HTML output.

    Escapes: & < > " '

    Args:
        value: String to sanitize.

    Returns:
        HTML-safe string.

    Example:
        sanitize_html('<script>alert("xss")</script>')
        # Returns: '&lt;script&gt;alert(&quot;xss&quot;)&lt;/script&gt;'
    """
    output, code = call_function("sanitize_html", value)
    if code != 0:
        raise MainframeFunctionError("sanitize_html", "Sanitization failed", code)
    return output


def sanitize_sql(value: str) -> str:
    """
    Sanitize string for SQL (basic escaping).

    WARNING: Prefer parameterized queries in production.

    Args:
        value: String to sanitize.

    Returns:
        SQL-safe string.

    Example:
        sanitize_sql("O'Reilly")  # "O''Reilly"
    """
    output, code = call_function("sanitize_sql", value)
    if code != 0:
        raise MainframeFunctionError("sanitize_sql", "Sanitization failed", code)
    return output


def sanitize_filename(name: str, replacement: str = "_") -> str:
    """
    Sanitize filename by removing dangerous characters.

    Args:
        name: Filename to sanitize.
        replacement: Character to replace dangerous chars with.

    Returns:
        Safe filename.

    Example:
        sanitize_filename("file/name:test.txt")  # "file_name_test.txt"
    """
    output, code = call_function("sanitize_filename", name, replacement)
    if code != 0:
        raise MainframeFunctionError("sanitize_filename", "Sanitization failed", code)
    return output


def sanitize_shell_arg(value: str) -> str:
    """
    Sanitize value for safe shell argument.

    Args:
        value: Value to sanitize.

    Returns:
        Shell-safe string.

    Example:
        sanitize_shell_arg("hello world; rm -rf /")
        # Returns properly escaped string
    """
    output, code = call_function("sanitize_shell_arg", value)
    if code != 0:
        raise MainframeFunctionError("sanitize_shell_arg", "Sanitization failed", code)
    return output
